Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/person/ContactDiaryAddPersonFragment.kt

```diff
@@ -0,0 +1,122 @@
+import de.rki.coronawarnapp.contactdiary.util.hideKeyboard
+import de.rki.coronawarnapp.databinding.ContactDiaryAddPersonFragmentBinding
+import de.rki.coronawarnapp.util.DialogHelper
+import de.rki.coronawarnapp.util.di.AutoInject
+import de.rki.coronawarnapp.util.ui.observe2
+import de.rki.coronawarnapp.util.ui.popBackStack
+import de.rki.coronawarnapp.util.ui.viewBindingLazy
+import de.rki.coronawarnapp.util.viewmodel.CWAViewModelFactoryProvider
+import de.rki.coronawarnapp.util.viewmodel.cwaViewModelsAssisted
+import javax.inject.Inject
+
+class ContactDiaryAddPersonFragment : Fragment(R.layout.contact_diary_add_person_fragment),
+    AutoInject {
+
+    private val binding: ContactDiaryAddPersonFragmentBinding by viewBindingLazy()
+
+    @Inject lateinit var viewModelFactory: CWAViewModelFactoryProvider.Factory
+    private val viewModel: ContactDiaryAddPersonViewModel by cwaViewModelsAssisted(
+        factoryProducer = { viewModelFactory },
+        constructorCall = { factory, _ ->
+            factory as ContactDiaryAddPersonViewModel.Factory
+            factory.create(navArgs.addedAt)
+        }
+    )
+
+    private val navArgs: ContactDiaryAddPersonFragmentArgs by navArgs()
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        val person = navArgs.selectedPerson
+        if (person != null) {
+            binding.apply {
+                contactDiaryPersonNameEditText.setText(person.fullName)
+                contactDiaryPersonPhoneNumberEditText.setText(person.phoneNumber)
+                contactDiaryPersonEmailEditText.setText(person.emailAddress)
+                contactDiaryPersonDeleteButton.visibility = View.VISIBLE
+                contactDiaryPersonDeleteButton.setOnClickListener {
+                    DialogHelper.showDialog(deletePersonConfirmationDialog)
+                }
+                contactDiaryPersonSaveButton.setOnClickListener {
+                    it.hideKeyboard()
+                    viewModel.updatePerson(person)
+                }
+            }
+            viewModel.apply {
+                nameChanged(person.fullName)
+                person.phoneNumber?.let { phoneNumberChanged(it) }
+                person.emailAddress?.let { emailAddressChanged(it) }
+            }
+        } else {
+            binding.contactDiaryPersonDeleteButton.visibility = View.GONE
+            binding.contactDiaryPersonSaveButton.setOnClickListener {
+                it.hideKeyboard()
+                viewModel.addPerson()
+            }
+        }
+
+        binding.apply {
+            contactDiaryPersonNameEditText.focusAndShowKeyboard()
+
+            contactDiaryPersonCloseButton.setOnClickListener {
+                it.hideKeyboard()
+                viewModel.closePressed()
+            }
+            contactDiaryPersonNameEditText.doAfterTextChanged {
+                viewModel.nameChanged(it.toString())
```

Comment:
Why do we need to update name/phone/email in the VM when the user switches the TextField? Isn't it enough to simply pass all data when the user clicks on Save?

Reply:
The save button is only enabled when you enter something in the name field. It's more of a failsafe so the user doesn't add empty entries in the db.